File: 우박수열 정적분.js
const makeArr = (n) => {
  const arr = [];
  while (n > 1) {
    arr.push(n);
    if (n % 2 === 0) {
      n /= 2;
    } else {
      n = n * 3 + 1;
    }
  }
  return [...arr, 1];
};

const solution = (k, ranges) => {
  const arr = makeArr(k);
  console.log(arr);

  return ranges.map((range) => {
    let start = range[0];
    let end = arr.length - 1 + range[1];

    if (start > end) {
      return -1.0;
    } else if (start === end) {
      return 0.0;
    }

    let sum = 0;
    while (start < end) {
      sum += (arr[start] + arr[start + 1]) / 2;
      start++;
    }
    return sum;
  });
};

console.log(
  solution(5, [
    [0, 0],
    [0, -1],
    [2, -3],
    [3, -3],
  ])
);

File: 구현/우박수열 정적분.py
# https://school.programmers.co.kr/learn/courses/30/lessons/134239
def get_area(a,b,arr):
    sum = 0
    for i in range(a,b):
        sum += (arr[i]+arr[i+1])/2
    return sum

def solution(k, ranges):
    answer = []
    arr = [k]
    cnt = 0
    while k != 1:
        if k % 2 == 0:
            k /= 2
        else:
            k = k*3+1
        cnt += 1
        arr.append(k)
    
    for a,b in ranges:
        b += cnt
        if a>b:
            answer.append(-1.0)
            continue
        answer.append(get_area(a,b,arr))
        
    return answer
